src/compiler/checker.ts

```diff
@@ -22168,8 +22168,15 @@ namespace ts {
         // constituent types keyed by the literal types of the property by that name in each constituent type.
         function getKeyPropertyName(unionType: UnionType): __String | undefined {
             const types = unionType.types;
-            // We only construct maps for large unions with non-primitive constituents.
-            if (types.length < 10 || getObjectFlags(unionType) & ObjectFlags.PrimitiveUnion) {
+            // We only construct maps for unions with many non-primitive constituents.
+            let nonPrimitiveCount = 0;
+            const isLargeNonPrimitives = !(getObjectFlags(unionType) & ObjectFlags.PrimitiveUnion) && forEach(types, t => {

```

Comment:
Consider retaining `types.length >= 10` as a way to avoid the `forEach` in easy cases.

Reply:
I also have slight preference for a plain loop, but I could go either way.

Reply:
without `forEach()` it'll be something like this:

```ts
            // We only construct maps for unions with many non-primitive constituents.
            if (types.length < 10 || getObjectFlags(unionType) & ObjectFlags.PrimitiveUnion) {
                return undefined;
            }
            let nonPrimitiveCount = 0;
            for (const t of types) {
                if ((t.flags & (TypeFlags.Primitive | TypeFlags.AnyOrUnknown | TypeFlags.Never))) {
                    continue;
                }
                nonPrimitiveCount++;
                if (nonPrimitiveCount >= 10) {
                    return undefined;
                }
            }
            if (nonPrimitiveCount >= 10) {
                return undefined;
            }
```